File: Topics/Arrays&Hashing/product_of_array_except_self/sln.cpp
// @Runtime: 15 ms (98.91%)
// @Memory: 23.9 MB (86.61%)

std::vector<int> productExceptSelf(std::vector<int> &input)
{
    std::vector<int> answer(input.size(), 1);

    for (int i = 0; i < input.size() - 1; ++i)
        answer[i + 1] = input[i] * answer[i];

    for (int i = input.size() - 1, product = input[i]; i > 0; --i, product *= input[i])
        answer[i - 1] *= product;

    return answer;
}

File: Topics/Arrays&Hashing/product_of_array_except_self/sln.rs
// @Runtime: 8 ms (84%)
// @Memory: 3.1 MB (94.29%)

pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
    let mut ans: Vec<i32> = vec![1; nums.len()];
    let mut prod: i32 = 1;

    for i in 1..nums.len() {
        ans[i] = nums[i - 1] * ans[i - 1];
    }

    for i in (0..nums.len() - 1).rev() {
        prod *= nums[i + 1];
        ans[i] *= prod;
    }

    ans
}
